File: examples/ir-spec/mod01.ml
type 'A my_pair2 = ('A, 'A) my_pair

and ('A, 'B) my_pair = 'A * 'B

let rec id'1 : 'X -> 'X = function v_X -> v_X

let rec arg13'3 : 'X * _ * _ -> 'X = function v_X, v_Y, v_Z -> v_X

let rec arg23'3 : _ * 'Y * _ -> 'Y = function v_X, v_Y, v_Z -> v_Y

let rec arg33'3 : _ * _ * 'Z -> 'Z = function v_X, v_Y, v_Z -> v_Z

let rec mk_int'0 : unit -> int = function () -> 1

let rec mk_float'0 : unit -> int = function () -> 2

let rec mk_string'0 : unit -> string = function () -> "Erlang String"

let rec mk_char'0 : unit -> char = function () -> 'c'

let rec mk_nil'0 : unit -> _ list = function () -> []

let rec mk_cons'2 : 'A * 'A list -> 'A list = function v_H, v_T -> v_H :: v_T

let rec mk_tuple'3 : 'A * 'B * 'C -> 'A * 'B * 'C = function
  | v_A, v_B, v_C -> (v_A, v_B, v_C)

let rec mk_tuple2'1 : 'A -> int * int * int * 'A = function
  | v_A -> (1, 2, 3, v_A)

let rec mk_map'2 :
    'A * 'B ->
    (< get_a : 'A ; set_a : 'A -> 'map_1 ; get_b : 'B ; set_b : 'B -> 'map_1 >
     as
     'map_1) = function
  | v_A, v_B ->
      object
        val val_a = v_A

        method get_a = val_a

        method set_a new_val_a = {<val_a = new_val_a>}

        val val_b = v_B

        method get_b = val_b

        method set_b new_val_b = {<val_b = new_val_b>}
      end

let rec update_map1'2 :
    (< get_a : 'A ; set_a : 'A -> 'map_2 > as 'map_2) * 'A ->
    (< get_a : 'A ; set_a : 'A -> 'map_3 > as 'map_3) = function
  | v_M, v_A ->
      let _ = Ffi.same'2 (v_M, v_M#set_a v_A) in
      let _ = Ffi.same'2 (v_A, v_M#get_a) in
      v_M#set_a v_A

let rec update_map2'2 :
    (< get_a : 'A ; set_a : 'A -> 'map_M ; .. > as 'map_M) * 'A ->
    (< get_a : 'A ; set_a : 'A -> 'map_M ; .. > as 'map_M) = function
  | v_M, v_A ->
      let _ = Ffi.same'2 (v_M, v_M#set_a v_A) in
      let _ = Ffi.same'2 (v_A, v_M#get_a) in
      v_M#set_a v_A

let rec update_map3'1 :
    (< get_a : 'A ; set_a : 'A -> 'map_4 ; get_b : 'A ; set_b : 'A -> 'map_4 >
     as
     'map_4) ->
    (< get_a : 'A ; set_a : 'A -> 'map_5 ; get_b : 'A ; set_b : 'A -> 'map_5 >
     as
     'map_5) = function
  | v_M ->
      let v_A0 = v_M#get_a in
      let v_B0 = v_M#get_a in
      let _ = Ffi.same'2 (v_M, v_M#set_a v_B0) in
      let _ = Ffi.same'2 (v_M, v_M#set_b v_A0) in
      let _ = Ffi.same'2 (v_B0, v_M#get_a) in
      let _ = Ffi.same'2 (v_A0, v_M#get_b) in
      (v_M#set_a v_B0)#set_b v_A0

let rec update_map4'1 :
    (< get_a : 'A
     ; set_a : 'A -> 'map_M
     ; get_b : 'A
     ; set_b : 'A -> 'map_M
     ; .. >
     as
     'map_M) ->
    (< get_a : 'A
     ; set_a : 'A -> 'map_M
     ; get_b : 'A
     ; set_b : 'A -> 'map_M
     ; .. >
     as
     'map_M) = function
  | v_M ->
      let v_A0 = v_M#get_a in
      let v_B0 = v_M#get_a in
      let _ = Ffi.same'2 (v_M, v_M#set_a v_B0) in
      let _ = Ffi.same'2 (v_M, v_M#set_b v_A0) in
      let _ = Ffi.same'2 (v_B0, v_M#get_a) in
      let _ = Ffi.same'2 (v_A0, v_M#get_b) in
      (v_M#set_a v_B0)#set_b v_A0

let rec access_map'1 :
    (< get_id : 'Id
     ; set_id : 'Id -> 'map__
     ; get_location : 'Location
     ; set_location : 'Location -> 'map__
     ; .. >
     as
     'map__) ->
    'Id * 'Location = function
  | v_M -> (v_M#get_id, v_M#get_location)

let rec access_map2'1 :
    (< get_inner1 : (< get_inner2 : 'A ; set_inner2 : 'A -> 'map_7 > as 'map_7)
     ; set_inner1 :
         (< get_inner2 : 'A ; set_inner2 : 'A -> 'map_7 > as 'map_7) -> 'map_6 >
     as
     'map_6) ->
    'A = function
  | v_M -> v_M#get_inner1#get_inner2

let rec mk_seq'0 : unit -> int * int = function
  | () ->
      let v_X = 1 in
      let v_Y = 2 in
      let _ = 3 in
      let _ = 4 in
      (v_X, v_Y)

let rec is_empty'1 : _ list -> bool = function
  | [] -> true
  | v_H :: v_T -> false

let rec is_empty2'1 : _ list -> bool = function [] -> true | _ -> false

let rec with_as'1 : int -> int = function 1 as v_X -> v_X | _ as v_Y -> v_Y

let rec block'2 : 'A * 'A list -> 'A list * 'A list = function
  | v_X, v_Y ->
      let v_Z = v_X :: v_Y in
      (v_Z, v_X :: v_Y)

let rec is_empty_case'1 : _ list -> bool = function
  | v_L -> ( match v_L with [] -> true | _ -> false )

let rec both_empty'2 : _ list * _ list -> bool = function
  | v_L1, v_L2 -> (
      match v_L1 with
      | [] -> ( match v_L2 with [] -> true | _ -> false )
      | _ -> false )

let rec call'2 : _ list * _ list -> bool = function
  | v_L1, v_L2 -> both_empty'2 (v_L1, v_L2)

let rec remote_call'1 : 'A list -> 'A list = function v_L -> List.rev v_L

let rec fun_to_var'0 : unit -> 'A * 'B -> 'A * 'B = function
  | () ->
      let v_F = function v_X, v_Y -> (v_X, v_Y) in
      v_F

let rec local_fun_to_var'2 : _ * _ -> _ list * _ list -> bool = function
  | v_A, v_B ->
      let v_F = call'2 in
      v_F

let rec remote_fun_to_var'2 : _ * _ -> 'A list -> 'A list = function
  | v_A, v_B ->
      let v_F = List.rev in
      v_F

let rec local_n_fun'0 : unit -> _ list -> _ list = function
  | () ->
      let v_F =
        let rec v_Local = function _ :: v_T -> v_Local v_T | [] -> [] in
        v_Local
      in
      v_F

let rec mod01F'1 : 'X -> 'X = function v_X -> v_X

let rec unary_plus'1 : int -> int = function v_X -> + (+v_X)

let rec unary_minus'1 : int -> int = function v_X -> - (-v_X)

let rec unary_not'1 : bool -> bool = function v_X -> not (not v_X)

let rec unary_bnot'1 : int -> int = function v_X -> lnot (lnot v_X)

let rec binary_star'2 : int * int -> int = function v_X, v_Y -> v_X * v_Y

let rec binary_div'2 : int * int -> int = function v_X, v_Y -> v_X / v_Y

let rec binary_rem'2 : int * int -> int = function v_X, v_Y -> v_X mod v_Y

let rec binary_band'2 : int * int -> int = function v_X, v_Y -> v_X land v_Y

let rec binary_and'2 : bool * bool -> bool = function v_X, v_Y -> v_X && v_Y

let rec binary_plus'2 : int * int -> int = function v_X, v_Y -> v_X + v_Y

let rec binary_minus'2 : int * int -> int = function v_X, v_Y -> v_X - v_Y

let rec binary_bor'2 : int * int -> int = function v_X, v_Y -> v_X lor v_Y

let rec binary_bxor'2 : int * int -> int = function v_X, v_Y -> v_X lxor v_Y

let rec binary_bsl'2 : int * int -> int = function v_X, v_Y -> v_X lsl v_Y

let rec binary_bsr'2 : int * int -> int = function v_X, v_Y -> v_X lsr v_Y

let rec binary_or'2 : bool * bool -> bool = function v_X, v_Y -> v_X || v_Y

let rec binary_xor'2 : bool * bool -> bool = function
  | v_X, v_Y -> (not v_X) <> not v_Y

let rec binary_orelse'2 : bool * bool -> bool = function
  | v_X, v_Y -> v_X || v_Y

let rec binary_andalso'2 : bool * bool -> bool = function
  | v_X, v_Y -> v_X && v_Y

let rec list_plus'2 : 'A list * 'A list -> 'A list = function
  | v_X, v_Y -> v_X @ v_Y

let rec list_minus'2 : 'A list * 'A list -> 'A list = function
  | v_X, v_Y -> Ffi.list_diff'2 (v_X, v_Y)

let rec comp1'2 : 'A * 'A -> bool = function v_X, v_Y -> v_X = v_Y

let rec comp2'2 : 'A * 'A -> bool = function v_X, v_Y -> v_X <> v_Y

let rec comp3'2 : 'A * 'A -> bool = function v_X, v_Y -> v_X <= v_Y

let rec comp4'2 : 'A * 'A -> bool = function v_X, v_Y -> v_X < v_Y

let rec comp5'2 : 'A * 'A -> bool = function v_X, v_Y -> v_X >= v_Y

let rec comp6'2 : 'A * 'A -> bool = function v_X, v_Y -> v_X > v_Y

let rec comp7'2 : 'A * 'A -> bool = function v_X, v_Y -> v_X = v_Y

let rec comp8'2 : 'A * 'A -> bool = function v_X, v_Y -> v_X <> v_Y

let rec guard1'2 : int * int -> bool = function
  | v_X, v_Y when (v_X = 1 && v_Y = 1) || v_X <> v_Y -> true
  | v_X, v_Y when false -> false
  | v_X, v_Y -> false

let rec guard2'3 : 'A list * 'A * 'A list -> 'A list = function
  | v_X, v_Y, [] -> (
      match v_X with v_X1 :: v_Xs when v_X1 = v_Y -> v_Xs | v_Xs -> v_Xs )
  | v_X, _, v_Z when v_X = v_Z -> v_Z
  | _, _, v_Z -> v_Z

let rec p_match_tuple0'1 : unit -> unit Ffi.tuple1 = function
  | () -> Ffi.Tuple1 ()

let rec p_match_tuple1'1 : _ Ffi.tuple1 -> unit = function
  | Ffi.Tuple1 v_X -> ()

let rec p_match_invoke'0 : unit -> unit = function
  | () ->
      let _ = p_match_tuple0'1 () in
      p_match_tuple1'1 (Ffi.Tuple1 ())

let rec any_id'1 : Ffi.any -> Ffi.any = function v_A -> v_A

let rec atom_id'1 : Ffi.atom -> Ffi.atom = function v_A -> v_A

let rec binary_id'1 : Ffi.binary -> Ffi.binary = function v_A -> v_A

let rec bitstring_id'1 : Ffi.bitstring -> Ffi.bitstring = function v_A -> v_A

let rec byte_id'1 : Ffi.byte -> Ffi.byte = function v_A -> v_A

let rec float_id'1 : float -> float = function v_X -> v_X

let rec identifier_id'1 : Ffi.identifier -> Ffi.identifier = function
  | v_A -> v_A

let rec iodata_id'1 : Ffi.iodata -> Ffi.iodata = function v_A -> v_A

let rec iolist_id'1 : Ffi.iolist -> Ffi.iolist = function v_A -> v_A

let rec map_id'1 : ('A, 'B) Ffi.map -> ('A, 'B) Ffi.map = function v_M -> v_M

let rec none_id'1 : Ffi.none -> Ffi.none = function v_A -> v_A

let rec noreturn_id'1 : Ffi.no_return -> Ffi.no_return = function
  | v_A -> none_id'1 v_A

let rec number_id'1 : Ffi.none -> Ffi.none = function v_A -> v_A

let rec pid_id'1 : Ffi.pid -> Ffi.pid = function v_A -> v_A

let rec port_id'1 : Ffi.port -> Ffi.port = function v_A -> v_A

let rec reference_id'1 : Ffi.reference -> Ffi.reference = function v_A -> v_A

let rec term_id'1 : Ffi.term -> Ffi.term = function v_A -> v_A

let rec timeout_id'1 : Ffi.timeout -> Ffi.timeout = function v_A -> v_A

let rec ints_id'1 :
    Ffi.neg_integer * Ffi.non_neg_integer * Ffi.pos_integer ->
    Ffi.neg_integer * Ffi.non_neg_integer * Ffi.pos_integer = function
  | v_X -> v_X

let rec mk_my_pair'2 = function v_A, v_B -> (v_A, v_B)

File: tests/dev/ir-spec/mod01.erl
-file("dev/src/mod01.erlt", 1).

-module(mod01).

-eqwalizer_unchecked([]).

-export_type([my_pair2/1, my_pair/2]).

-export([mod01F/1,
         id/1,
         arg13/3,
         arg23/3,
         arg33/3,
         mk_int/0,
         mk_float/0,
         mk_string/0,
         mk_char/0,
         mk_nil/0,
         mk_cons/2,
         mk_tuple/3,
         mk_tuple2/1,
         mk_shape/2,
         update_shape1/2,
         update_shape2/2,
         update_shape3/1,
         update_shape4/1,
         access_shape/1,
         access_shape2/1,
         mk_seq/0,
         is_empty/1,
         is_empty2/1,
         with_as/1,
         block/2,
         is_empty_case/1,
         both_empty/2,
         call/2,
         fun_to_var/0,
         local_fun_to_var/2,
         local_n_fun/0,
         unary_plus/1,
         unary_minus/1,
         unary_not/1,
         unary_bnot/1,
         binary_star/2,
         binary_div/2,
         binary_rem/2,
         binary_band/2,
         binary_and/2,
         binary_plus/2,
         binary_minus/2,
         binary_bor/2,
         binary_bxor/2,
         binary_bsl/2,
         binary_bsr/2,
         binary_or/2,
         binary_xor/2,
         binary_orelse/2,
         binary_andalso/2,
         list_plus/2,
         list_minus/2,
         comp1/2,
         comp2/2,
         comp3/2,
         comp4/2,
         comp5/2,
         comp6/2,
         comp7/2,
         comp8/2,
         guard1/2,
         guard2/3,
         p_match_tuple0/1,
         p_match_tuple1/1,
         p_match_invoke/0,
         any_id/1,
         atom_id/1,
         binary_id/1,
         bitstring_id/1,
         byte_id/1,
         float_id/1,
         identifier_id/1,
         iodata_id/1,
         iolist_id/1,
         none_id/1,
         noreturn_id/1,
         number_id/1,
         pid_id/1,
         port_id/1,
         reference_id/1,
         term_id/1,
         timeout_id/1,
         ints_id/1,
         mk_my_pair/2]).

-spec id(Arg1 :: X) -> X.

id(X) -> X.

-spec arg13(Arg1 :: X, _, _) -> X.

arg13(X, _Y, _Z) -> X.

-spec arg23(_, Arg2 :: Y, _) -> Y.

arg23(_X, Y, _Z) -> Y.

-spec arg33(_, _, Arg3 :: Z) -> Z.

arg33(_X, _Y, Z) -> Z.

-spec mk_int() -> integer().

mk_int() -> 1.

-spec mk_float() -> integer().

mk_float() -> 2.12312309999999992982.

-spec mk_string() -> string().

mk_string() -> "Erlang String".

-spec mk_char() -> char().

mk_char() -> $c.

-spec mk_nil() -> [_].

mk_nil() -> [].

-spec mk_cons(H :: A, T :: [A]) -> [A].

mk_cons(H, T) -> [H | T].

-spec mk_tuple(A, B, C) -> {A, B, C}.

mk_tuple(A, B, C) -> {A, B, C}.

-spec mk_tuple2(A) -> {El1 :: integer(),
                       El2 :: integer(),
                       integer(),
                       El4 :: A}.

mk_tuple2(A) -> {1, 2, 3, A}.

-spec mk_shape(A, B) -> #{a := A, b := B}.

mk_shape(A, B) -> #{a => A, b => B}.

-spec update_shape1(#{a := A}, A) -> #{a := A}.

update_shape1(S, A) -> S#{a => A}.

-spec update_shape2(#{a := A, atom() => any()},
                    A) -> #{a := A, atom() => any()}.

update_shape2(S, A) -> S#{a => A}.

-spec update_shape3(#{a := A, b := A}) -> #{a := A,
                                            b := A}.

update_shape3(S) ->
    A0 = erlang:map_get(a, S),
    B0 = erlang:map_get(a, S),
    S#{a => B0, b => A0}.

-spec update_shape4(#{a := A, b := A,
                      atom() => any()}) -> #{a := A, b := A, atom() => any()}.

update_shape4(S) ->
    A0 = erlang:map_get(a, S),
    B0 = erlang:map_get(a, S),
    S#{a => B0, b => A0}.

-spec access_shape(#{id := Id, location := Location,
                     atom() => any()}) -> {Id, Location}.

access_shape(S) ->
    {erlang:map_get(id, S), erlang:map_get(location, S)}.

-spec access_shape2(#{inner1 := #{inner2 := A}}) -> A.

access_shape2(S) ->
    erlang:map_get(inner2, erlang:map_get(inner1, S)).

-spec mk_seq() -> {integer(), integer()}.

mk_seq() ->
    X = 1,
    Y = 2,
    {X, Y}.

-spec is_empty([_]) -> boolean().

is_empty([]) -> true;
is_empty([_H | _T]) -> false.

-spec is_empty2([_]) -> boolean().

is_empty2([]) -> true;
is_empty2(_) -> false.

-spec with_as(integer()) -> integer().

with_as(1 = X) -> X;
with_as(_ = Y) -> Y.

-spec block(A, [A]) -> {[A], [A]}.

block(X, Y) ->
    {begin Z = [X | Y], Z end, begin [X | Y] end}.

-spec is_empty_case([_]) -> boolean().

is_empty_case(L) ->
    case L of
        [] -> true;
        _ -> false
    end.

-spec both_empty([_], [_]) -> boolean().

both_empty(L1, L2) ->
    case L1 of
        [] ->
            case L2 of
                [] -> true;
                _ -> false
            end;
        _ -> false
    end.

-spec call([_], [_]) -> boolean().

call(L1, L2) -> mod01:both_empty(L1, L2).

-spec fun_to_var() -> fun((A, B) -> {A, B}).

fun_to_var() ->
    F = fun (X, Y) -> {X, Y} end,
    F.

-spec local_fun_to_var(_, _) -> fun(([_],
                                     [_]) -> boolean()).

local_fun_to_var(_A, _B) ->
    F = fun call/2,
    F.

-spec local_n_fun() -> fun(([_]) -> [_]).

local_n_fun() ->
    F = fun Local([_ | T]) -> Local(T);
            Local([]) -> []
        end,
    F.

-spec mod01F(X) -> X.

mod01F(X) -> X.

-spec unary_plus(integer()) -> integer().

unary_plus(X) -> +(+X).

-spec unary_minus(integer()) -> integer().

unary_minus(X) -> -(-X).

-spec unary_not(boolean()) -> boolean().

unary_not(X) -> not (not X).

-spec unary_bnot(integer()) -> integer().

unary_bnot(X) -> bnot (bnot X).

-spec binary_star(integer(), integer()) -> integer().

binary_star(X, Y) -> X * Y.

-spec binary_div(integer(), integer()) -> integer().

binary_div(X, Y) -> X div Y.

-spec binary_rem(integer(), integer()) -> integer().

binary_rem(X, Y) -> X rem Y.

-spec binary_band(integer(), integer()) -> integer().

binary_band(X, Y) -> X band Y.

-spec binary_and(boolean(), boolean()) -> boolean().

binary_and(X, Y) -> X and Y.

-spec binary_plus(integer(), integer()) -> integer().

binary_plus(X, Y) -> X + Y.

-spec binary_minus(integer(), integer()) -> integer().

binary_minus(X, Y) -> X - Y.

-spec binary_bor(integer(), integer()) -> integer().

binary_bor(X, Y) -> X bor Y.

-spec binary_bxor(integer(), integer()) -> integer().

binary_bxor(X, Y) -> X bxor Y.

-spec binary_bsl(integer(), integer()) -> integer().

binary_bsl(X, Y) -> X bsl Y.

-spec binary_bsr(integer(), integer()) -> integer().

binary_bsr(X, Y) -> X bsr Y.

-spec binary_or(boolean(), boolean()) -> boolean().

binary_or(X, Y) -> X or Y.

-spec binary_xor(boolean(), boolean()) -> boolean().

binary_xor(X, Y) -> X xor Y.

-spec binary_orelse(boolean(), boolean()) -> boolean().

binary_orelse(X, Y) -> X orelse Y.

-spec binary_andalso(boolean(), boolean()) -> boolean().

binary_andalso(X, Y) -> X andalso Y.

-spec list_plus([A], [A]) -> [A].

list_plus(X, Y) -> X ++ Y.

-spec list_minus([A], [A]) -> [A].

list_minus(X, Y) -> X -- Y.

-spec comp1(A, A) -> boolean().

comp1(X, Y) -> X == Y.

-spec comp2(A, A) -> boolean().

comp2(X, Y) -> X /= Y.

-spec comp3(A, A) -> boolean().

comp3(X, Y) -> X =< Y.

-spec comp4(A, A) -> boolean().

comp4(X, Y) -> X < Y.

-spec comp5(A, A) -> boolean().

comp5(X, Y) -> X >= Y.

-spec comp6(A, A) -> boolean().

comp6(X, Y) -> X > Y.

-spec comp7(A, A) -> boolean().

comp7(X, Y) -> X =:= Y.

-spec comp8(A, A) -> boolean().

comp8(X, Y) -> X =/= Y.

-spec guard1(integer(), integer()) -> boolean().

guard1(X, Y) when X == 1, Y == 1; X =/= Y -> true;
guard1(X, Y) when X > Y -> false;
guard1(_X, _Y) -> false.

-spec guard2([A], A, [A]) -> [A].

guard2(X, Y, []) ->
    case X of
        [X1 | Xs] when X1 == Y -> Xs;
        Xs -> Xs
    end;
guard2(X, _, Z) when X == Z -> Z;
guard2(_, _, Z) -> Z.

-spec p_match_tuple0({}) -> {{}}.

p_match_tuple0({}) -> {{}}.

-spec p_match_tuple1({_}) -> {}.

p_match_tuple1({_X}) -> {}.

-spec p_match_invoke() -> {}.

p_match_invoke() ->
    p_match_tuple0({}),
    p_match_tuple1({{}}).

-spec any_id(any()) -> any().

any_id(A) -> A.

-spec atom_id(atom()) -> atom().

atom_id(A) -> A.

-spec binary_id(binary()) -> binary().

binary_id(A) -> A.

-spec bitstring_id(bitstring()) -> bitstring().

bitstring_id(A) -> A.

-spec byte_id(byte()) -> byte().

byte_id(A) -> A.

-spec float_id(float()) -> float().

float_id(X) -> X.

-spec identifier_id(identifier()) -> identifier().

identifier_id(A) -> A.

-spec iodata_id(iodata()) -> iodata().

iodata_id(A) -> A.

-spec iolist_id(iolist()) -> iolist().

iolist_id(A) -> A.

-spec none_id(none()) -> none().

none_id(A) -> A.

-spec noreturn_id(no_return()) -> no_return().

noreturn_id(A) -> none_id(A).

-spec number_id(number()) -> number().

number_id(A) -> A.

-spec pid_id(pid()) -> pid().

pid_id(A) -> A.

-spec port_id(port()) -> port().

port_id(A) -> A.

-spec reference_id(reference()) -> reference().

reference_id(A) -> A.

-spec term_id(term()) -> term().

term_id(A) -> A.

-spec timeout_id(timeout()) -> timeout().

timeout_id(A) -> A.

-spec ints_id({neg_integer(),
               non_neg_integer(),
               pos_integer()}) -> {neg_integer(),
                                   non_neg_integer(),
                                   pos_integer()}.

ints_id(X) -> X.

-type my_pair2(A) :: mod01:my_pair(A, A).

-type my_pair(A, B) :: {A, B}.

-spec mk_my_pair(A, A) -> mod01:my_pair2(A).

mk_my_pair(A, B) -> {A, B}.



